x/cdp/keeper/interest.go

```diff
@@ -161,11 +162,111 @@ func (k Keeper) CalculateNewInterest(ctx sdk.Context, cdp types.CDP) sdk.Coin {
 }
 
 // SynchronizeInterestForRiskyCDPs synchronizes the interest for the slice of cdps with the lowest collateral:debt ratio
-func (k Keeper) SynchronizeInterestForRiskyCDPs(ctx sdk.Context, slice sdkmath.Int, targetRatio sdk.Dec, collateralType string) error {
-	cdps := k.GetSliceOfCDPsByRatioAndType(ctx, slice, targetRatio, collateralType)
-	for _, cdp := range cdps {
+func (k Keeper) SynchronizeInterestForRiskyCDPs(ctx sdk.Context, targetRatio sdk.Dec, cp types.CollateralParam) error {
+	debtParam := k.GetParams(ctx).DebtParam
+
+	cdpStore := prefix.NewStore(ctx.KVStore(k.key), types.CdpKeyPrefix)
+	collateralRatioStore := prefix.NewStore(ctx.KVStore(k.key), types.CollateralRatioIndexPrefix)
+
+	cdpIDs := make([]uint64, 0, cp.CheckCollateralizationIndexCount.Int64())
+
+	iterator := collateralRatioStore.Iterator(types.CollateralRatioIterKey(cp.Type, sdk.ZeroDec()), types.CollateralRatioIterKey(cp.Type, targetRatio))
+	for ; iterator.Valid(); iterator.Next() {
+		_, id, _ := types.SplitCollateralRatioKey(iterator.Key())
+		cdpIDs = append(cdpIDs, id)
+		if int64(len(cdpIDs)) >= cp.CheckCollateralizationIndexCount.Int64() {
+			break
+		}
+	}
+	iterator.Close()
+
+	globalInterestFactor, found := k.GetInterestFactor(ctx, cp.Type)
+	if !found {
+		globalInterestFactor = sdk.OneDec()
+		k.SetInterestFactor(ctx, cp.Type, globalInterestFactor)
```

Comment:
@rhuairahrighairidh returning here might be better? Though, I don't believe there is a case where a cdp can exist without a global interest factor set (create cdp sets it), so regardless believe it to be compatible with old behavior

Reply:
yeah slight lean to just returning. I don't think it matters but not having interest sync (from a early return) feels not as bad as potentially incorrectly setting a global factor